File: server/routes/consultants.js
const express = require('express');
const axios = require('axios');

const router = express.Router();

const AIRTABLE_API_KEY = process.env.AIRTABLE_API_KEY;
const AIRTABLE_BASE_ID = process.env.AIRTABLE_BASE_ID;
const AIRTABLE_TABLE_NAME = process.env.AIRTABLE_TABLE_NAME;
const AIRTABLE_VIEW_ID = process.env.AIRTABLE_VIEW_ID;

function mapRecord(record) {
  const f = record.fields;
  return {
    id: record.id,
    name: f['Full Legal Name'] || '—',
    employeeNumber: f['Employee Number'] || '—',
    employeeId: f['Employee ID'] || '—',
    idNumber: f['ID Number'] || '—',
    client: f['Client Name'] || '—',
    contractStart: f['Contract Start Date'] || null,
    contractEnd: f['Contract End Date'] || null,
    salary: f['Current Monthly Salary'] || null,
    bankName: f['Bank Name'] || '—',
    accountNo: f['Bank Account Number'] || '—',
  };
}

// Fetch all pages from Airtable (100 records per page)
async function fetchAllRecords() {
  const records = [];
  let offset = null;

  do {
    const params = {
      view: AIRTABLE_VIEW_ID,
      pageSize: 100,
    };
    if (offset) params.offset = offset;

    const res = await axios.get(
      `https://api.airtable.com/v0/${AIRTABLE_BASE_ID}/${encodeURIComponent(AIRTABLE_TABLE_NAME)}`,
      {
        headers: { Authorization: `Bearer ${AIRTABLE_API_KEY}` },
        params,
      }
    );

    records.push(...res.data.records);
    offset = res.data.offset || null;
  } while (offset);

  return records;
}

router.get('/', async (req, res) => {
  if (!AIRTABLE_API_KEY || !AIRTABLE_BASE_ID || !AIRTABLE_TABLE_NAME) {
    return res.status(503).json({ error: 'Airtable not configured' });
  }

  try {
    const raw = await fetchAllRecords();
    const consultants = raw.map(mapRecord);
    res.json({ consultants, total: consultants.length });
  } catch (err) {
    console.error('Airtable fetch error:', err.response?.data || err.message);
    res.status(502).json({ error: 'Failed to fetch consultant data from Airtable' });
  }
});

module.exports = router;

File: app/routers/consultants.py
import httpx
from urllib.parse import quote
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from app.config import AIRTABLE_API_KEY, AIRTABLE_BASE_ID, AIRTABLE_TABLE_NAME, AIRTABLE_VIEW_ID

router = APIRouter()


def _map_record(record: dict) -> dict:
    f = record.get("fields", {})
    salary_raw = f.get("Current Monthly Salary")
    salary = None
    if salary_raw is not None:
        try:
            salary = float(str(salary_raw).replace(",", ""))
        except (ValueError, TypeError):
            salary = None
    return {
        "id": record["id"],
        "name": f.get("Full Legal Name") or "-",
        "employeeNumber": f.get("Employee Number") or "-",
        "employeeId": f.get("Employee ID") or "-",
        "idNumber": f.get("ID Number") or "-",
        "client": f.get("Client Name") or "-",
        "contractStart": f.get("Contract Start Date"),
        "contractEnd": f.get("Contract End Date"),
        "salary": salary,
        "bankName": f.get("Bank Name") or "-",
        "accountNo": f.get("Bank Account Number") or "-",
    }


@router.get("/api/consultants")
async def get_consultants(request: Request):
    if not AIRTABLE_API_KEY or not AIRTABLE_BASE_ID or not AIRTABLE_TABLE_NAME:
        return JSONResponse({"error": "Airtable not configured"}, status_code=503)
    try:
        AIRTABLE_API_KEY.encode("ascii")
    except UnicodeEncodeError:
        return JSONResponse({"error": "AIRTABLE_API_KEY contains invalid characters — check Vercel env vars"}, status_code=503)

    records = []
    offset = None
    try:
        async with httpx.AsyncClient(timeout=30) as client:
            while True:
                params = {
                    "pageSize": 100,
                    "cellFormat": "string",
                    "timeZone": "Asia/Kuala_Lumpur",
                    "userLocale": "en-MY",
                }
                if AIRTABLE_VIEW_ID:
                    params["view"] = AIRTABLE_VIEW_ID
                if offset:
                    params["offset"] = offset
                url = (
                    f"https://api.airtable.com/v0/"
                    f"{quote(AIRTABLE_BASE_ID, safe='')}/{quote(AIRTABLE_TABLE_NAME, safe='')}"
                )
                resp = await client.get(
                    url,
                    headers={"Authorization": f"Bearer {AIRTABLE_API_KEY}"},
                    params=params,
                )
                if not resp.is_success:
                    return JSONResponse({"error": f"Airtable error {resp.status_code}"}, status_code=502)
                data = resp.json()
                records.extend(data.get("records", []))
                offset = data.get("offset")
                if not offset:
                    break
    except Exception as e:
        return JSONResponse({"error": str(e)}, status_code=502)

    consultants = [_map_record(r) for r in records]
    return JSONResponse({"consultants": consultants, "total": len(consultants)})
